python/ray/autoscaler/v2/instance_manager/subscribers/ray_stopper.py

```diff
@@ -0,0 +1,66 @@
+import logging
+from concurrent.futures import ThreadPoolExecutor
+from typing import List
+
+from ray._raylet import GcsClient
+from ray.autoscaler.v2.instance_manager.instance_manager import (
+    InstanceUpdatedSubscriber,
+)
+from ray.core.generated.autoscaler_pb2 import DrainNodeReason
+from ray.core.generated.instance_manager_pb2 import (
+    Instance,
+    InstanceUpdateEvent,
+    TerminationRequest,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class RayStopper(InstanceUpdatedSubscriber):
+    """RayStopper is responsible for stopping ray on instances."""
+
+    def __init__(self, gcs_client: GcsClient) -> None:
+        self._gcs_client = gcs_client
+        self._executor = ThreadPoolExecutor(max_workers=1)
+
+    def notify(self, events: List[InstanceUpdateEvent]) -> None:
+        for event in events:
+            if event.new_instance_status == Instance.RAY_STOPPING:
+                self._executor.submit(self._stop_ray, event)
+
+    def _stop_ray(self, event: InstanceUpdateEvent) -> None:
+        termination_request = event.termination_request
+        ray_node_id = termination_request.ray_node_id
+
+        if termination_request.cause == TerminationRequest.Cause.IDLE:
+            reason = DrainNodeReason.DRAIN_NODE_REASON_IDLE_TERMINATION
+            reason_str = "Idle termination of node for {} seconds.".format(
+                termination_request.idle_time_ms / 1000
+            )
+        elif (
+            termination_request.cause == TerminationRequest.Cause.MAX_NUM_NODE_PER_TYPE
+        ):
+            reason = DrainNodeReason.DRAIN_NODE_REASON_PREEMPTION
+            reason_str = "Preempted due to max number of nodes per type={}.".format(
+                termination_request.max_num_nodes_per_type
+            )
+        elif termination_request.cause == TerminationRequest.Cause.MAX_NUM_NODES:
+            reason = DrainNodeReason.DRAIN_NODE_REASON_PREEMPTION
+            reason_str = "Preempted due to max number of nodes={}.".format(
+                termination_request.max_num_nodes
+            )
+
+        reply = self._gcs_client.drain_node(
```

Comment:
a blocking rpc call seems dangerous to me.

If it will timeout inside gcs_client automatically, could you add comment here?

gRPC could hang with random reason until the connection timeout reached.

Reply:
Ah, i see, do we have an async version for the gcs client? The GcsAioClient?

I think relying on the gcs's timeout is fine here. 
- The method is ran in an executor so if we block here, the caller wouldn't be affected 
- We had timeout mechanism to reconcile cases where we got stuck at draining ray node (we would shut down the underlying cloud node in this case)